library/proc_macro/src/quote.rs

```diff
@@ -73,7 +75,7 @@ pub fn quote(stream: TokenStream) -> TokenStream {
                 after_dollar = false;
                 match tree {
                     TokenTree::Ident(_) => {
-                        return Some(quote!(Into::<crate::TokenStream>::into(
+                        return Some(quote!(ToTokens::into_token_stream(
                         Clone::clone(&(@ tree))),));
```

Comment:
I think the way this function works should be changed; instead of having a big tree that gets `.collect()`ed here, it should create a `TokenStream` at the beginning and then use `ToTokens::to_tokens` to append to it, rather than always cloning.

That should also let the loop be flattened a bit:

```rust
pub fn quote(stream: TokenStream) -> TokenStream {
    let mut output = TokenStream::new();
    if stream.is_empty() {
        return output;
    }
    let proc_macro_crate = quote!(crate);
    let mut after_dollar = false;

    for tree in stream {
        if after_dollar {
            after_dollar = false;
            match tree { /* ... */ }
            continue;
        }

        if let TokenTree::punct(ref tt) = tree {
            if tt.as_char() == '$' {
                after_dollar = true;
                continue;
            }
        }

        match tree {
            // similar to what exists except the arms will just use `to_tokens`
            // rather than cloning and returning something.
        }
    }
}
```

It may also be good to rename the `macro_rules! quote` in this file to `quote_impl` since reading all the `quote!` invocations in this function is somewhat misleading.

Reply:
+1 to both suggestions.

Starting with an empty `TokenStream::new()` and appending to it using the interpolated variable's `to_tokens` is how the `quote` crate's current expansion works.